File: app/schemas/account.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Generic, TypeVar

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.utils import ASSET_CLASS_DEPOSIT, ASSET_CLASS_STOCK

T = TypeVar("T")


class ORMModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int
    total_pages: int


class AccountTypeBrief(ORMModel):
    id: int
    code: str
    name: str
    category: str
    supports_holdings: bool


class AccountTypeResponse(ORMModel):
    id: int
    code: str
    name: str
    category: str
    supports_holdings: bool
    supports_contribution_limit: bool
    sort_order: int
    is_system: bool


class AccountTypeCreate(BaseModel):
    code: str
    name: str
    category: str
    supports_holdings: bool = False
    supports_contribution_limit: bool = False
    sort_order: int = 0


class AccountSummary(BaseModel):
    holdings_count: int
    holdings_value: Decimal
    total_value: Decimal


class AccountResponse(ORMModel):
    id: int
    account_type_id: int
    account_type: AccountTypeBrief | None = None
    name: str
    institution: str | None
    cash_balance: Decimal
    metadata: dict[str, Any] = Field(default_factory=dict)
    is_active: bool
    summary: AccountSummary | None = None


class AccountCreate(BaseModel):
    account_type_id: int
    name: str
    institution: str | None = None
    cash_balance: Decimal = Decimal("0")
    metadata: dict[str, Any] | None = None


class AccountUpdate(BaseModel):
    account_type_id: int | None = None
    name: str | None = None
    institution: str | None = None
    cash_balance: Decimal | None = None
    metadata: dict[str, Any] | None = None
    is_active: bool | None = None


class HoldingResponse(ORMModel):
    id: int
    account_id: int
    account_name: str | None = None
    asset_class: str
    symbol: str
    name: str
    quantity: Decimal
    avg_cost_price: Decimal
    manual_price: Decimal | None
    last_market_price: Decimal | None
    last_price_updated_at: datetime | None
    current_price: Decimal
    market_value: Decimal
    cost_basis: Decimal
    profit_loss: Decimal
    profit_loss_rate: Decimal
    interest_rate: Decimal | None = None
    start_date: date | None = None
    maturity_date: date | None = None
    accrued_interest: Decimal | None = None


class HoldingCreate(BaseModel):
    account_id: int
    asset_class: str = ASSET_CLASS_STOCK
    symbol: str | None = None
    name: str
    quantity: Decimal
    avg_cost_price: Decimal | None = None
    interest_rate: Decimal | None = None
    start_date: date | None = None
    maturity_date: date | None = None

    @model_validator(mode="after")
    def validate_deposit(self):
        if self.asset_class == ASSET_CLASS_DEPOSIT:
            if self.avg_cost_price is None:
                self.avg_cost_price = Decimal("1")
            return self
        if not self.symbol:
            raise ValueError("symbol이 필요합니다.")
        if self.avg_cost_price is None:
            raise ValueError("avg_cost_price가 필요합니다.")
        return self


class HoldingUpdate(BaseModel):
    asset_class: str | None = None
    symbol: str | None = None
    name: str | None = None
    quantity: Decimal | None = None
    avg_cost_price: Decimal | None = None
    manual_price: Decimal | None = None
    interest_rate: Decimal | None = None
    start_date: date | None = None
    maturity_date: date | None = None


class RefreshPricesRequest(BaseModel):
    holding_ids: list[int] | None = None


class RefreshPricesResponse(BaseModel):
    updated: int
    failed: list[dict[str, Any]]


class InvestmentTransactionResponse(ORMModel):
    id: int
    account_id: int
    holding_id: int | None
    type: str
    transaction_date: date
    quantity: Decimal | None
    price: Decimal | None
    amount: Decimal
    fee: Decimal
    memo: str | None


class InvestmentTransactionCreate(BaseModel):
    account_id: int
    holding_id: int | None = None
    asset_class: str | None = None
    symbol: str | None = None
    name: str | None = None
    type: str
    transaction_date: date
    quantity: Decimal | None = None
    price: Decimal | None = None
    amount: Decimal
    fee: Decimal = Decimal("0")
    memo: str | None = None
    sync_to_ledger: bool = False
    ledger_category_id: int | None = None


class InvestmentTransactionUpdate(BaseModel):
    type: str | None = None
    transaction_date: date | None = None
    quantity: Decimal | None = None
    price: Decimal | None = None
    amount: Decimal | None = None
    fee: Decimal | None = None
    memo: str | None = None


class AccountLimitResponse(ORMModel):
    id: int
    account_id: int
    account_name: str | None = None
    year: int
    contribution_limit: Decimal
    contributed_amount: Decimal
    remaining_amount: Decimal
    usage_rate: Decimal


class AccountLimitUpsert(BaseModel):
    account_id: int
    year: int
    contribution_limit: Decimal

File: src/lib/validations/account.ts
import Decimal from "decimal.js";
import { z } from "zod";

import { ASSET_CLASS_DEPOSIT, ASSET_CLASS_STOCK } from "@/lib/utils";

export const decimalSchema = z
  .union([z.string(), z.number(), z.instanceof(Decimal)])
  .transform((value) => new Decimal(value));

export const avgCostPriceSchema = decimalSchema.transform((value) =>
  value.toDecimalPlaces(4, Decimal.ROUND_HALF_UP),
);

export const bookCostSchema = decimalSchema.transform((value) =>
  value.toDecimalPlaces(0, Decimal.ROUND_HALF_UP),
);

export const dateSchema = z.preprocess(
  (value) => {
    if (value instanceof Date) {
      return value;
    }
    if (typeof value === "string" && /^\d{4}-\d{2}-\d{2}$/.test(value)) {
      return new Date(`${value}T00:00:00.000Z`);
    }
    return value;
  },
  z.coerce.date(),
);

export const paginatedResponseSchema = <T extends z.ZodType>(itemSchema: T) =>
  z.object({
    items: z.array(itemSchema),
    total: z.number().int(),
    page: z.number().int(),
    page_size: z.number().int(),
    total_pages: z.number().int(),
  });

export const accountTypeBriefSchema = z.object({
  id: z.number().int(),
  code: z.string(),
  name: z.string(),
  category: z.string(),
  supports_holdings: z.boolean(),
});

export const accountTypeResponseSchema = z.object({
  id: z.number().int(),
  code: z.string(),
  name: z.string(),
  category: z.string(),
  supports_holdings: z.boolean(),
  supports_contribution_limit: z.boolean(),
  sort_order: z.number().int(),
  is_system: z.boolean(),
});

export const accountTypeCreateSchema = z.object({
  code: z.string(),
  name: z.string(),
  category: z.string(),
  supports_holdings: z.boolean().default(false),
  supports_contribution_limit: z.boolean().default(false),
  sort_order: z.number().int().default(0),
});

export const accountSummarySchema = z.object({
  holdings_count: z.number().int(),
  holdings_value: decimalSchema,
  total_value: decimalSchema,
});

export const accountResponseSchema = z.object({
  id: z.number().int(),
  account_type_id: z.number().int(),
  account_type: accountTypeBriefSchema.nullable().optional(),
  name: z.string(),
  institution: z.string().nullable(),
  cash_balance: decimalSchema,
  metadata: z.record(z.string(), z.unknown()).default({}),
  is_active: z.boolean(),
  summary: accountSummarySchema.nullable().optional(),
});

export const accountCreateSchema = z.object({
  account_type_id: z.number().int(),
  name: z.string(),
  institution: z.string().nullable().optional(),
  cash_balance: decimalSchema.default(new Decimal(0)),
  metadata: z.record(z.string(), z.unknown()).nullable().optional(),
});

export const accountUpdateSchema = z.object({
  account_type_id: z.number().int().optional(),
  name: z.string().optional(),
  institution: z.string().nullable().optional(),
  cash_balance: decimalSchema.optional(),
  metadata: z.record(z.string(), z.unknown()).nullable().optional(),
  is_active: z.boolean().optional(),
});

export const holdingResponseSchema = z.object({
  id: z.number().int(),
  account_id: z.number().int(),
  account_name: z.string().nullable().optional(),
  asset_class: z.string(),
  symbol: z.string(),
  name: z.string(),
  quantity: decimalSchema,
  avg_cost_price: decimalSchema,
  manual_price: decimalSchema.nullable(),
  last_market_price: decimalSchema.nullable(),
  last_price_updated_at: z.coerce.date().nullable(),
  current_price: decimalSchema,
  market_value: decimalSchema,
  cost_basis: decimalSchema,
  profit_loss: decimalSchema,
  profit_loss_rate: decimalSchema,
  interest_rate: decimalSchema.nullable().optional(),
  start_date: dateSchema.nullable().optional(),
  maturity_date: dateSchema.nullable().optional(),
  accrued_interest: decimalSchema.nullable().optional(),
});

export const holdingCreateSchema = z
  .object({
    account_id: z.number().int(),
    asset_class: z.string().default(ASSET_CLASS_STOCK),
    symbol: z.string().nullable().optional(),
    name: z.string(),
    quantity: decimalSchema,
    avg_cost_price: avgCostPriceSchema.nullable().optional(),
    book_cost: bookCostSchema.nullable().optional(),
    interest_rate: decimalSchema.nullable().optional(),
    start_date: dateSchema.nullable().optional(),
    maturity_date: dateSchema.nullable().optional(),
  })
  .superRefine((data, ctx) => {
    if (data.asset_class === ASSET_CLASS_DEPOSIT) {
      return;
    }
    if (!data.symbol) {
      ctx.addIssue({ code: "custom", message: "symbol이 필요합니다.", path: ["symbol"] });
    }
    if (data.avg_cost_price == null) {
      ctx.addIssue({
        code: "custom",
        message: "avg_cost_price가 필요합니다.",
        path: ["avg_cost_price"],
      });
    }
  });

export const holdingUpdateSchema = z.object({
  asset_class: z.string().optional(),
  symbol: z.string().optional(),
  name: z.string().optional(),
  quantity: decimalSchema.optional(),
  avg_cost_price: avgCostPriceSchema.optional(),
  book_cost: bookCostSchema.nullable().optional(),
  manual_price: decimalSchema.nullable().optional(),
  interest_rate: decimalSchema.nullable().optional(),
  start_date: dateSchema.nullable().optional(),
  maturity_date: dateSchema.nullable().optional(),
});

export const refreshPricesRequestSchema = z.object({
  holding_ids: z.array(z.number().int()).nullable().optional(),
});

export const refreshPricesResponseSchema = z.object({
  updated: z.number().int(),
  failed: z.array(z.record(z.string(), z.unknown())),
});

export const investmentTransactionResponseSchema = z.object({
  id: z.number().int(),
  account_id: z.number().int(),
  holding_id: z.number().int().nullable(),
  holding_name: z.string().nullable().optional(),
  holding_symbol: z.string().nullable().optional(),
  type: z.string(),
  transaction_date: dateSchema,
  quantity: decimalSchema.nullable(),
  price: decimalSchema.nullable(),
  amount: decimalSchema,
  fee: decimalSchema,
  tax: decimalSchema,
  memo: z.string().nullable(),
});

export const investmentTransactionCreateSchema = z.object({
  account_id: z.number().int(),
  holding_id: z.number().int().nullable().optional(),
  asset_class: z.string().nullable().optional(),
  symbol: z.string().nullable().optional(),
  name: z.string().nullable().optional(),
  type: z.string(),
  transaction_date: dateSchema,
  quantity: decimalSchema.nullable().optional(),
  price: decimalSchema.nullable().optional(),
  amount: decimalSchema,
  fee: decimalSchema.default(new Decimal(0)),
  tax: decimalSchema.default(new Decimal(0)),
  memo: z.string().nullable().optional(),
  sync_to_ledger: z.boolean().default(false),
  ledger_category_id: z.number().int().nullable().optional(),
});

export const investmentTransactionUpdateSchema = z.object({
  type: z.string().optional(),
  transaction_date: dateSchema.optional(),
  quantity: decimalSchema.optional(),
  price: decimalSchema.optional(),
  amount: decimalSchema.optional(),
  fee: decimalSchema.optional(),
  tax: decimalSchema.optional(),
  memo: z.string().nullable().optional(),
});

export const accountLimitResponseSchema = z.object({
  id: z.number().int(),
  account_id: z.number().int(),
  account_name: z.string().nullable().optional(),
  year: z.number().int(),
  contribution_limit: decimalSchema,
  contributed_amount: decimalSchema,
  remaining_amount: decimalSchema,
  usage_rate: decimalSchema,
});

export const accountLimitUpsertSchema = z.object({
  account_id: z.number().int(),
  year: z.number().int(),
  contribution_limit: decimalSchema,
});

export type AccountTypeBrief = z.infer<typeof accountTypeBriefSchema>;
export type AccountResponse = z.infer<typeof accountResponseSchema>;
export type HoldingResponse = z.infer<typeof holdingResponseSchema>;
export type AccountSummary = z.infer<typeof accountSummarySchema>;
export type AccountLimitResponse = z.infer<typeof accountLimitResponseSchema>;
export type InvestmentTransactionResponse = z.infer<typeof investmentTransactionResponseSchema>;
